### PR TITLE
Feature : Filtering nodes by there domain name and distance

### LINKED ISSUE #16: Add filtering by nodes to the Linked Commons. (reference idealist)
## Problem Description
 Currently the visualization of the graph is not provide any insights due to large no. of nodes and edges plotted altogether. And also plotting whole graph will create a rendering process too slow.

## Solution Description 
Adding node filter will lead to analyse the particular node with better visualization. The main thing here to focus is that cc dealing with large no. of nodes and links so the time complexity of all the algorithms which are going to implemented must be optimize enough.

So, I came up with an algorithm which will filters the nodes by there domain name and going only  the particular distance as selected by the user with O(N+M) complexity where N,M are the no. of nodes and edges respectively.

1. Every time fetching data from the json file for every filter query is not feasible, so instead we will create graph data structure as user session starts and that data structure will going to use for further filter queries.
That will save time for fetching data every time from json file.

2. Now, I am using the Breadth First Search (BFS) which will traverse the graph level by level and terminate as it reached to the particular distance provided by user. In case distance parameter is missing at that time BFS will traverse till leaf nodes.

3. When the traversal of graph completed that data will passed to forced graph for plotting that's it.

## Snapshots

1. Filter query on domain with autocomplete feature.

CC_Output_2

2. Filter nodes by providing traversing depth/distance.

CC_Output_3

3. Final Plottings

CC_Output_1

CC_Output_4

### PR DESCRIPTION
**Pull request by mann2108**
Fixes #16

## Description

Implemented filter by nodes and distance successfully.
Refer _"Documentation.txt"_ file for reviewing changes that I made.

## Checklist
- [X] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [X] My pull request targets the `master` branch of the repository.
- [X] My commit messages follow [best practices][best_practices].
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [X] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

### CODE REVIEW

File: GSoC2019/src/index.html [hunk 1]
```diff
@@ -10,15 +11,38 @@
...
-   <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js" integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q" crossorigin="anonymous"></script>
-<script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js" integrity="sha384-JZR6Spejh4U02d8jOt6vLEHfe/JQGiRRSQQxSfFWpi1MquVdAyjUar5+76PVCmYl" crossorigin="anonymous"></script>
-<link rel="stylesheet" href="css/materialize.css" type="text/css">
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js" integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q" crossorigin="anonymous"></script>
+  <link rel="stylesheet" href="css/materialize.css" type="text/css">
   <!--<script src="../../dist/force-graph.js"></script>-->
+  
+  <!--Jquery-->
+  <link rel="stylesheet" href="https://code.jquery.com/ui/1.12.1/themes/smoothness/jquery-ui.css">  
+  <script src="https://code.jquery.com/jquery-3.4.1.js"></script>
```
Comment (mathemancer) [inline GSoC2019/src/index.html]: Is it necessary to pull in this version of `jquery`?  Note that there's already one stored in the `js` directory.

**Review by mathemancer — CHANGES_REQUESTED** [review]
~The initial problem I'm noticing is that something odd is happening with the text scaling when I filter.  For example, try filtering by 'flickr' with distance 3.~

Disregard this.  It appears to be a previously introduced bug.

**Comment by mathemancer** [comment]
image

**Comment by mathemancer** [comment]
More comments I've thought of:
1. Is it possible to do the distance filtering with the native library?
1. I think it'd be good to show the graph initially, before any filtering.
1. When one changes the distance, the `Plot Graph` button is kind of in the wrong place.  Perhaps the search box and drop down should be adjacent, and the `Plot Graph` button to the right.

File: GSoC2019/src/index.html (hunk 1 above)
Comment (mann2108) [reply GSoC2019/src/index.html]: Yeah. Its not necessary to pull this version for now. I will remove this in my later version.

**Comment by mann2108** [comment]
> More comments I've thought of:
> 
> 1. Is it possible to do the distance filtering with the native library?
> 2. I think it'd be good to show the graph initially, before any filtering.
> 3. When one changes the distance, the `Plot Graph` button is kind of in the wrong place.  Perhaps the search box and drop down should be adjacent, and the `Plot Graph` button to the right.



> More comments I've thought of:
> 
> 1. Is it possible to do the distance filtering with the native library?
> 2. I think it'd be good to show the graph initially, before any filtering.
> 3. When one changes the distance, the `Plot Graph` button is kind of in the wrong place.  Perhaps the search box and drop down should be adjacent, and the `Plot Graph` button to the right.

**Comment by mann2108** [comment]
> More comments I've thought of:
> 
> 1. Is it possible to do the distance filtering with the native library?

I can't find a way to implement this with native library. I will try to figure out the implementation  with native library if possible.

> 2. I think it'd be good to show the graph initially, before any filtering.
> 3. When one changes the distance, the `Plot Graph` button is kind of in the wrong place.  Perhaps the search box and drop down should be adjacent, and the `Plot Graph` button to the right.

Both the above points seem good to me too. I will update it.

**Sorry I closed this issue by mistake.**

**Comment by mann2108** [comment]
@mathemancer  by mistake i have removed the request for creativecommons/community-cc-catalog-data-viz. Please have a look into this.

**Comment by mann2108** [comment]
@mathemancer I have successfully committed the changes. You can review it and let me know the further updates.

**Comment by soccerdroid** [comment]
Hi @mann2108 , sorry for the late comment, we were figuring out the next steps for the project. A couple of observations:

- For screens smaller than 550px, the page has to be responsive, and some elements would disappear. Actually what we plan to do is to leave interactions with the graph only for larger screens (tablets and so on). The search bar and distance field should change positions and sizes for mobile phones.
- @subhamX just made changes in the front-end, and also integrated react to the project. I suggest you check it (the way the search bar and distance field look like is different now). 
- I will reorganize this repository, so the project in production will be the one with react. The GSoC2019 folder will be archived. I realize this affects your pr, so please take my apologies. 

Please, for now check the new code @subhamX has added. One thing we are sorting out right now is text searching (among all the domain names, not just the ones of the nodes we are rendering). Ideas and contributions are welcome here.

**Comment by mann2108** [comment]
Hello @soccerdroid @subhamX ,

I went through the react version of the project. I have noticed one point that will help us to achieve text searching among all nodes instead of those we last queried. 

The class "Graph2D" have one object "linksPerDomain" that will updated after each filter query and also we are using the same object for parsing the filter query in "handleFilterSubmit" method (That's why may be we able to see the alert of "Invalid Node Name" for the correct domain name). To solve this we need another object which may be constant throughout the all queries which initialize with "componentDidMount()" method and that will be used for all other queries just like “OriginalGraphData”.
And this task can be achieved with an alternate approach too like fetching data every time from the json file but this approach is not feasible for large no. of nodes.

And for the responsive design we can go for the react-bootstrap module. Where we can replace the current navbar with bootstrap version which automatically updates the position of elements as per the screen resolution.
reference link - https://react-bootstrap.github.io/layout/grid/

@subhamX let me know your reviews.

**Comment by subhamX** [comment]
@mann2108 I think you intend to improve the current state of autocomplete feature. Actually, whatever you are seeing right now is just the browser's default behaviour for `input` element and nothing extra is added to tweak the behaviour.

**Regarding the search issue**
Yes, you are right that for some nodes like **twitter** etc you will get the **Invalid Node Name** prompt. This is because of the way **linksPerDomains** (which is kind of adjacency list) is built/implemented. To be precise, if you type any node which has an **outdegree of 0** then it will lead to that Invalid Query prompt (as there won't be any instance in the linksPerDomains).

**Why this isn't fixed yet?**
Actually, we have moved the filtering logic from client-side to server-side and the feature is already implemented. The case mentioned above is taken care too. Source code is in `core-server` branch. Work is almost finished and the PR is in the review stage.

Coming to auto-complete feature. We wish to have a more scalable system and want to reduce the client overhead. For this reason, we are considering to have the autocomplete logic on the server-side. We are still looking for all options and do share your views on the server-side autocomplete feature. Any ideas will be a great help!

Temporary link of the [latest deployed version](https://ccdataviz.netlify.app/). Do share your general thoughts too! :)

**Comment by mann2108** [comment]
@subhamX sure, will get back to you soon.
Currently I am seeing this front end which is different from the deployed version.
image

**Comment by subhamX** [comment]
@mann2108  let me share the branch details in brief, which should give a fair idea of which component is in which branch.

**react-migration:** Refactors and migrate the project to react with client-side filtering
**core-server:** Add Django server which is responsible for server-side filtering
**design-upgrade:** Redesign the project which more resembles to creative commons aesthetics.

Both `core-server` and `design-upgrade` are development branches whereas `react-migration` is already merged with `master`. The latest frontend version is in `design-upgrade` and the latest deployed server version is in `core-server` branch.

**Comment by mann2108** [comment]
@subhamX Thanks for clarifying that. :+1:

**Comment by soccerdroid** [comment]
Closing this pr due to inactivity

**Comment by soccerdroid** [comment]
Closing this pr due to inactivity

**Pull request closed without merging**